### ISSUE TITLE
Inconsistent behavior when changing number of devices.

### ISSUE THREAD

**Issue by dirkweissenborn** [facebookresearch/detr#354]
Changing the number of devices leads to change of per device batch size. This in turn leads to change in training behavior because the classification loss is normalized per device while the bbox losses are normalized by the global number of boxes. Since the per device batch size is very small, this can have quite a big affect on training.

In a reimplementation we found that fixing this issue leads to a significant drop in performance, so we would assume that this would also be the case for this implementation.

**Comment by alcinos** [comment]
Hi @dirkweissenborn,
Thanks for reaching out.

Could you clarify with what batch size and what batch size/device you are experimenting? On our side here are generally our observations with respect to keeping the global batch size constant while changing the batch size/device:
 - With bs=1/device, some quirky behaviour may occur at inference if one try to evaluate with a bs >1/device. The reason behind this is that with 1 image per card, padding due to different aspect ratio never occurs, hence the model never learns to handle it. If padding occurs at test time, it negatively impacts the predictions.
 - bs=2/device is what we use by default
 - bs=4/device works almost the same (up to negligible perf variation).
 - bs>4 we have not experimented with that as we hit the gpu memory limit at some point (and it's quite slow anyway)

Hope this helps.

**Comment by dirkweissenborn** [comment]
This was just a general observation as we reimplemented detr. The problem
is that the normalization of the ce loss is local, i.e., per device, and
that of the box losses is global. That makes the loss depend on the number
of devices which is not a good thing to have and might lead to unexpected
behavior. The variance of the denominator in the ce loss (which is implicit
in the cross entropy) is quite high with smallish batch size per device.
This is due to the fact that the number of objects between examples varies
quite a bit.

Note I am not sure that this really leads to varying performance, but just
wanted to bring up the observation, as it seems like a bug to me. An easy
way out would be to do the normalization of the ce loss manually as in the
box loss case.
